### PR TITLE
SDK-399 - signing tool for enclave purposes

### PR DESCRIPTION
**Pull request by i-skrypnyk**

### CODE REVIEW

**Review by paolocappelletti — CHANGES_REQUESTED** [review]

**Review by cronicc — CHANGES_REQUESTED** [review]

File: tools/signingtool/src/main/java/com/horizen/SigningToolCommandProcessor.java [hunk 1]
```diff
@@ -0,0 +1,311 @@
...
+                        "\tverifySignature <arguments>\n" +
+                        "\tsignMessage <arguments>\n" +
+                        "\tvalidateMessage <arguments>\n" +
+                        "\tprivKeyToPubkey <arguments>\n" +
+                        "\texit\n"
+        );
+    }
+
+    private void printCreateSignatureUsageMsg(String error) {
+        printer.print("Error: " + error);
```
[the remark this answers is not in the document]
Comment (i-skrypnyk) [reply tools/signingtool/src/main/java/com/horizen/SigningToolCommandProcessor.java:135]: Hi @cronicc ! These are like command line style messages, I can turn them into a json like these:
```
{
"Usage" : [
                "From command line: <program name> <command name> [<json data>]",
                "For interactive mode: <command name> [<json data>]",
                "Read command arguments from file: <command name> -f <path to file with json data>" 
        ],
"Supported commands" : [
                "help",
                "createSignature <arguments>",
                "verifySignature <arguments>",
                "signMessage <arguments>",
                "validateMessage <arguments>",
                "privKeyToPubKey <arguments>",
                "exit"
        ]
}
```
and
```
{
        "Error": "error",
        "Usage": "createSignature { \"message\": message, \"privateKey\": private_key, \"type\": \"string\" [\"schnorr\"] } "
}
```

Does that look fine to you?

**Review by paolocappelletti — APPROVED** [review]

**Review by tvangundy-hl — APPROVED** [review]

File: tools/signingtool/src/main/java/com/horizen/SigningToolCommandProcessor.java [hunk 2]
```diff
@@ -0,0 +1,310 @@
...
+        byte[] prefix = json.get("prefix").asText().getBytes();
+        byte[] prefixMessage = hash(concatenate(prefix, message));
+
+        verifyMessage(signature, publicKey, prefixMessage);
+    }
+
+    private void printPrivKeyToPubKeyUsageMsg(String error) {
+        ObjectNode resJson = new ObjectMapper().createObjectNode();
+
+        resJson.put("Error", error);
```
Comment (i-Alex) [inline tools/signingtool/src/main/java/com/horizen/SigningToolCommandProcessor.java:235]: according to the specs: `error` must be lowercase, not `Error`.
Let's change it to keep namings in the same style.
Here and in other places

File: tools/signingtool/src/main/java/com/horizen/SigningToolCommandProcessor.java [hunk 3]
```diff
@@ -0,0 +1,310 @@
...
+            printPrivKeyToPubKeyUsageMsg("privateKey is not specified or has invalid format.");
+            return;
+        } else if (!json.has("type") || !json.get("type").isTextual() || !Objects.equals(json.get("type").asText(), "schnorr")) {
+            printPrivKeyToPubKeyUsageMsg("type is not specified or has invalid format.");
+            return;
+        }
+
+        String privateKey = json.get("privateKey").asText();
+        Secret secret = SECRETS_COMPANION.parseBytes(BytesUtils.fromHexString(privateKey));
+        assert secret.secretTypeId() == SecretsIdsEnum.SchnorrSecretKeyId.id();
```
Comment (i-Alex) [inline tools/signingtool/src/main/java/com/horizen/SigningToolCommandProcessor.java:252]: It seems that `assert` usage is too much here.
Much better to return a readable error: "Provided privateKey type is not supported...."
So the caller will be able to see the clear error message.
We should avoid processing errors in `processCommand` method.

Same in other places.

File: tools/signingtool/src/main/java/com/horizen/SigningToolCommandProcessor.java [hunk 4]
```diff
@@ -0,0 +1,310 @@
...
+        if (!json.has("privateKey") || !json.get("privateKey").isTextual()) {
+            printPrivKeyToPubKeyUsageMsg("privateKey is not specified or has invalid format.");
+            return;
+        } else if (!json.has("type") || !json.get("type").isTextual() || !Objects.equals(json.get("type").asText(), "schnorr")) {
+            printPrivKeyToPubKeyUsageMsg("type is not specified or has invalid format.");
+            return;
+        }
+
+        String privateKey = json.get("privateKey").asText();
+        Secret secret = SECRETS_COMPANION.parseBytes(BytesUtils.fromHexString(privateKey));
```
Comment (i-Alex) [inline tools/signingtool/src/main/java/com/horizen/SigningToolCommandProcessor.java:251]: wrap with try catch and return a human readable `error` message to the caller.
Currently, if someone will path a random string -> IllegalArgumentException will be thrown and not processed and processed in the method invoker as a generic error.

File: tools/signingtool/src/test/java/com/horizen/SigningToolTest.java [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+        argsJson = new ObjectMapper().createObjectNode();
+        argsJson.put("message", testMessage);
+        argsJson.put("signature", signature);
+        argsJson.put("publicKey", BytesUtils.toHexString(PROPOSITION_SERIALIZER.toBytes(publicKey)));
+        argsJson.put("type", SCHNORR);
+        args = argsJson.toString();
+        result = runTest(new String[]{command, args});
+        assertTrue(result != null && result.contains("valid") && result.contains("true"));
+
+        //Negative
```
Comment (i-Alex) [inline tools/signingtool/src/test/java/com/horizen/SigningToolTest.java:90]: add tests with invalid signature and invalid public keys. Also signature or key of invalid types (that can't be even deserialized)
Same in other tests

**Review by i-Alex — CHANGES_REQUESTED** [review]
Looks good.
I can see that we need to process parsing errors with more context in the error messages.
Some negative test cases should be added.

Another point to discuss, is to put the placeholder for different keys/signature schemes support.
I'm pretty sure that quite soon we will add other core key types.

File: tools/signingtool/src/test/java/com/horizen/SigningToolTest.java [hunk 6]
```diff
@@ -0,0 +1,224 @@
...
+        argsJson = new ObjectMapper().createObjectNode();
+        argsJson.put("privateKey", BytesUtils.toHexString(SECRETS_COMPANION.toBytes(secretKey)) + "123");
+        argsJson.put("type", SCHNORR);
+        args = argsJson.toString();
+
+        result = runTest(new String[]{PRIV_KEY_TO_PUB_KEY, args});
+        assertTrue(result != null && result.contains("error") && result.contains("Unable to parse privateKey"));
+    }
+
+}
```
Comment (cronicc) [inline tools/signingtool/src/test/java/com/horizen/SigningToolTest.java:224]: @i-skrypnyk when running inside a container without networking (like in the secure enclave):
```
java -jar tools/signingtool/target/sidechains-sdk-signingtools-0.5.0-SNAPSHOT.jar
2022-10-28 10:34:11,405 main ERROR Could not determine local host name java.net.UnknownHostException: 95c76d23dcac: 95c76d23dcac: Temporary failure in name resolution
        at java.base/java.net.InetAddress.getLocalHost(Unknown Source)
        at org.apache.logging.log4j.core.util.NetUtils.getLocalHostname(NetUtils.java:54)
        at org.apache.logging.log4j.core.LoggerContext.lambda$setConfiguration$0(LoggerContext.java:620)
        at java.base/java.util.concurrent.ConcurrentHashMap.computeIfAbsent(Unknown Source)
        at org.apache.logging.log4j.core.LoggerContext.setConfiguration(LoggerContext.java:620)
        at org.apache.logging.log4j.core.LoggerContext.reconfigure(LoggerContext.java:699)
        at org.apache.logging.log4j.core.LoggerContext.reconfigure(LoggerContext.java:716)
        at org.apache.logging.log4j.core.LoggerContext.start(LoggerContext.java:270)
        at org.apache.logging.log4j.core.impl.Log4jContextFactory.getContext(Log4jContextFactory.java:155)
        at org.apache.logging.log4j.core.impl.Log4jContextFactory.getContext(Log4jContextFactory.java:47)
        at org.apache.logging.log4j.LogManager.getContext(LogManager.java:196)
        at org.apache.logging.log4j.LogManager.getLogger(LogManager.java:599)
        at com.horizen.SigningTool.main(SigningTool.java:27)
Caused by: java.net.UnknownHostException: 95c76d23dcac: Temporary failure in name resolution
        at java.base/java.net.Inet4AddressImpl.lookupAllHostAddr(Native Method)
        at java.base/java.net.InetAddress$PlatformNameService.lookupAllHostAddr(Unknown Source)
        at java.base/java.net.InetAddress.getAddressesFromNameService(Unknown Source)
        at java.base/java.net.InetAddress$NameServiceAddresses.get(Unknown Source)
        at java.base/java.net.InetAddress.getAllByName0(Unknown Source)
        ... 13 more

{"Usage":["From command line: <program name> <command name> [<json data>]"],"Supported commands":["help","createSignature <arguments>","verifySignature <arguments>","signMessage <arguments>","validateMessage <arguments>","privKeyToPubKey <arguments>"]}
```

**Review by cronicc — CHANGES_REQUESTED** [review]

File: tools/signingtool/src/test/java/com/horizen/SigningToolTest.java (hunk 6 above)
Comment (i-skrypnyk) [reply tools/signingtool/src/test/java/com/horizen/SigningToolTest.java:224]: @cronicc Hi, can you please check with the latest fix, I removed the logging completely for the signingtool, and in my tests log4j is no longer initialized, thus not invoking InetAddress.getLocalHost. But I don't know how to correctly emulate your environment.

**Review by i-Alex — APPROVED** [review]
LGTM!

**Review by cronicc — APPROVED** [review]

File: tools/signingtool/src/test/java/com/horizen/SigningToolTest.java (hunk 6 above)
Comment (cronicc) [reply tools/signingtool/src/test/java/com/horizen/SigningToolTest.java:224]: @i-skrypnyk error is gone now. To replicate `docker run --rm -v $(pwd)/tools/signingtool/target/:/target --network=none eclipse-temurin:11-jre-jammy java -jar /target/sidechains-sdk-signingtools-0.5.0-SNAPSHOT.jar`

**Pull request merged**